dynamicdns

steglicd opened the pull request:
Would be nice if freedombox could be reached via DNS resolved hostname (even if freedombox is connected via dynamic IP).

For this reason I'm implementing a ez-ipupdate module.

The module should detect the current WAN IP (even if freedombox is connected behind NAT router) and push the IP to a open source GnudIP server if a change is detected.

IPv6 and other Server types than gnudip should also be included later.

GnudIP vs. update URL:

* Gnudip is more safe (md5 hashed and salted password transmission prevent replay attacks)
* update URL is more generic (the most common dynamic DNS Service provides a URL to update the IP)

steglicd commented:
some things needs to be done:
* currently I used the icon from XMPP server for plinth UI menu entry 
* use Django FormView to eliminate some boilerplate.
* add ez-ipupdate to freedombox image (is freedom-maker on github?)
* use package framework
* some further testing (if someone would support testing I can provide gnudip accounts)

jvalleroy commented:
freedom-maker is here: https://github.com/freedombox/freedom-maker
But the default list of packages is actually in freedombox-setup: https://github.com/petterreinholdtsen/freedombox-setup/blob/master/debian/control

steglicd commented:
Thanks for this info, according to README of freedombox-setup:
"Note that most setup should be done as plinth modules on request of its user, and not during installation."

so there should be a "install" button on module page instead of deliver it within the freedombox image?

SunilMohanAdapa commented:
On Monday 05 January 2015 01:04 AM, steglicd wrote:
[...]
>   * add ez-ipupdate to freedombox image (is freedom-maker on github?)
>   * use package framework
>   * some further testing (if someone would support testing I can provide
>     gnudip accounts)

On Monday 05 January 2015 01:28 AM, James Valleroy wrote:
> freedom-maker is here: https://github.com/freedombox/freedom-maker
> But the default list of packages is actually in freedombox-setup:
>
https://github.com/petterreinholdtsen/freedombox-setup/blob/master/debian/control

One of the things about the package framework is that it installs the
necessary packages when needed.  This is on-demand install approach has
following advantages.

- If we run the package as daemon, it will take up resources even for
people who don't need it.  If we have small amount of RAM or large
number of such applications, we are in trouble.

- If we adopt the on-demand model, life will be easier for application
developers.  They would not have to modify and rebuild freedombox
images, but ship an extra FreedomBox app!  This will considerably
improve our application development time.

- FreedomBox images will be very lean.  Even for people installing
freedombox-setup on regular machines, things will be better.

- Debugging problems with install from Plinth are much more easier than
debugging issues with freedombox-setup.

Due to above reasons, I believe that we should have most packages
install (except essentials like firewall) in on-demand mode.  We should
even remove some of the current packages like pagekite from
freedombox-setup dependencies list.

-- 
Sunil

steglicd commented:
OK, so I will wait for package framework.

SunilMohanAdapa commented:
On Monday 05 January 2015 02:37 AM, steglicd wrote:
> OK, so I will wait for package framework.

It is now available on master.

-- 
Sunil

steglicd commented:
Ok, thanks. I pushed my changes. Package framework is used now.

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns.html:
```diff
@@ -0,0 +1,28 @@
+#
+{% endcomment %}
+
+{% load bootstrap %}
```
The `load bootstrap` provides the `bootstrap` template filters (which you don't use in this file and in dynamicdns_status.html)

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_status.html:
```diff
@@ -0,0 +1,42 @@
+  <p>
+  <h3>NAT type</h3>
+  {% if nat_unchecked %}
+  	NAT type not detected yet, if you do not provide a "IP check URL" we will 
```
There are spaces and tabs mixed for indentation here. Depending on the editor this should be pretty easy to change, for vim it's just `:retab`.

So far we use 2 spaces to indent html and javascript (as you are doing, except for the tabs)

fonfon commented:
It would be good to propose the user some free dyndns servers, or to post a link to a list.
I just tried https://www.noip.com which works with ez-ipupdate, but still the Plinth DynamicDNS configuration page times hangs and times out when I enter the configuration details. Tell me if you can't reproduce that, then I'll have a closer look.

The python files are not pep8, you could use a code checking tool like python3-flake8 (easy) or pylint (more complete but more complicated).

steglicd commented:
I do not have a noip account (do they offer free accounts? I could not find a free offer). could you please run:

sh -x /usr/share/plinth/actions/dynamicdns configure -s gnudip.datasystems24.net -d fbx.sds-ip.de -u fbx -p frdm -I http://myip.datasystems24.de

This should work. Next time you could use your noip account settings and try again. Please post the output in case of error.

which files are not pep8? when I run "pep8 *.py" within my module folder I do not get any errors.
The python3-flake8 tool you are using comes from which package?

fonfon commented:
python3-flake8 is the debian unstable package name.

    fonfon@freedombox:~/code/plinth/plinth/modules/dynamicdns$ flake8 *.py
    dynamicdns.py:45:80: E501 line too long (80 > 79 characters)
    dynamicdns.py:216:9: E125 continuation line with same indent as next logical line
    dynamicdns.py:245:39: E502 the backslash is redundant between brackets
    dynamicdns.py:246:13: E128 continuation line under-indented for visual indent
    dynamicdns.py:256:33: E502 the backslash is redundant between brackets
    dynamicdns.py:257:9: E128 continuation line under-indented for visual indent

noip.com offers free accounts but you have to renew/reactivate them every 30 days, I just registered to test the pull request. I'll have a look at the configuration now.

fonfon commented:
I found three issues so far:

1. the `actions/dynamicdns` fails on line 94: 

    rm: cannot remove ‘/etc/ez-ipupdate/ez-ipupdate.status’: No such file or directory

2. the form help text of the password field is misleading:
"You should have been requested to select a password when you created the account. If you left this field empty your password will be unchanged."
But in https://github.com/steglicd/Plinth/blob/dynamicdns/plinth/modules/dynamicdns/dynamicdns.py#L216 you require the password to be re-entered otherwise the changes won't take effect.
I'd recommend using django forms like i.e. in `modules/first_boot/forms.py`.

3. the timeout/hanging issue seems to come from https://github.com/steglicd/Plinth/blob/dynamicdns/plinth/modules/dynamicdns/dynamicdns.py#L245
Perhaps stopping or configuring isn't finished when start is run and so it fails? If I step through the process slowly everything works fine.

steglicd commented:
@fonfon 
In L216 I check if there was a password configured previously. So if you configure the module 1st time, you are forced to enter a password. When you change any other parameter later you don't need to re-enter the password again.

But you are right, the password checking magic is quite brain fucking - If django Forms supply a pasword checking function it will be way better to use this instead.

BobGirard commented:
Given the understandable confusion about how to do the code checking, it seems all the more reasonable to resolve issue #58 ("Add code checking tools"), which I in fact volunteered to address.  However, there seemed to be some unresolved disagreements about (1) whether there would be any specific implementation (that is, via a shell script or custom command) or simply dependencies on the appropriate tools, and (2) which specific tool(s) to use (i.e., flake8 vs. pylint).  Accordingly, I've held off, awaiting a more explicit consensus.  So do we have a clearer idea now as to how to proceed?

steglicd commented:
The issue "web page is hanging because of unresponsive script in case of dynamic DNS Server does not support gnudip" should be fixed. I also added a link to get a free dynamic DNS account and removed some helpe text which causes confusion. Still need to refactore the password input boxes.

fonfon commented:
@steglicd : Do you still plan to change the password input boxes (switching to django forms I assume), or should I review the pull request as it is right now?

steglicd commented:
@fonfon 
I already used django forms (like: TrimmedCharField( .... , widget=forms.PasswordInput() ...).
What I was looking for is a better way to handle the following requirements:
* show a password box and a password confirmation box
* compare the input of this two fields, raise an error if the input does not match
* force the user to fill both boxes (required=True) BUT: only if no password was set before (i.e. if the user only want to change "IP check URL" he should NOT be forced to enter the password again). Raise an error if the Password Box is empty AND no password was set before.

I have implemented this behaviour by myself but I was looking for a django way of doing stuff like this (maybe some extended password object which will do this). Maybe you can give me a hint?

fonfon commented:
This can be solved with django form validation. Basically django offers field validation (for one input field) and a form validation for the whole form. As you have two input fields which are sometimes optional and depend on each other, the form validation sounds like the most appropriate way: https://docs.djangoproject.com/en/1.7/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other

fonfon commented:
About making the password fields optional sometimes:
You could perhaps solve this by changing the field attributes in the `__init__` method of the form. If that's not possible in a straightforward way I'd just raise a `ValidationError` inside of the `clean` method I mentioned before, this is probably the easier way and works for sure.

steglicd commented:
Ok, I will have a look into this and tell you when I'm done.

BTW:
there are 2 more tasks left:
1) a GUI Icon
2) even if gnudip is the most secure way of dynamic DNS update, many DynamicDNS Providers are not supporting gnudip. So a generic HTTP(s) "update URL" way of doing the update should be added. 

Advantage:
* should work with most providers

Disadvantages:
* The user has to find out the URL (like https://dynamic-dns.tld/update.php?User=Username&Pass=Password&Domain=domain&IP=ip)
* The user needs to handle variables (maybe not "end-user-compatible"), because parts of the URL are variable. i.e. he need to add the update URL like: "https://dynamic-dns.tld/update.php?User=$Username&Pass=$Password&Domain=$domain&IP=$ip"
* Replay attack is easy on Update URL's (so I may capture your domain)

What do you think?

fonfon commented:
at 1): Currently we just use the most suitable bootstrap glyphicon: http://getbootstrap.com/components/#glyphicons-glyphs
at 2): my opinion: If the configuration via URL allows to use many dynamic DNS providers that wouldn't work otherwise then it's worth the disadvantages. Perhaps we can provide pre-defined URLs for the most common dynamic DNS providers? (If you use JavaScript take care that the basic functionality also works with JavaScript disabled)

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns.html:
```diff
@@ -0,0 +1,26 @@
+
+{% block content %}
+<h2>DynamicDNS client</h2>
+<p>If your internet provider changes your IP address periodic (i.e. every 24h) it may be hard for others to find you in the WEB. And for this reason nobody may find the services which are provided by FreedomBox (like your ownCloud). The solution is to assign a DNS name to your IP address and update the DNS name every time your IP is changed by your Internet provider. Dynamic DNS allows you to pus ...
```
some more linebreaks (<80 characters) would be nice to improve the readability

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,266 @@
+                             _('Dynamic DNS configuration is updated!'))
+    else:
+        messages.error(request,
+                       _('At least on failure occured,\
```
typo: one

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,266 @@
+        messages.error(request, _('please give a password'))
+        fail = True
+
+    if False == fail:
```
I'd find a `if not fail:` better readable

steglicd commented:
@fonfon thanks for the review. I made the changes as you suggested. 
I have one more question left. I made form validation within the clean() function (check if password is empty, if only one field is filled or if the passwords differ).

But if I want to check if the user configured a password before, I need the data from "old_status" array. Within _apply_changes I have this information. Do you know how to get this from within the clear() function (without calling get_status() again). I wan't to avoid calling get_status() again because this may need time to run (and the information should be available yet).  Currently this check is still done within _apply_changes but I would like to move this to the FormValidation process.

steglicd commented:
PS.: update URL is on my ToDo List

fonfon commented:
About the two password fields: 
Is it really necessary to have two password fields? I'd say two password fields are to make sure a user doesn't have a typo when he chooses a new password. But in this case the password is created elsewhere so there is no need to deal with this problem.
There might even be a 'show password' checkbox that displays the password in plain text.

About accessing password fields in `clean`:
You can provide the status data to the form when creating it, using the `initial` parameter. In this case it's on https://github.com/steglicd/Plinth/blob/dynamicdns/plinth/modules/dynamicdns/dynamicdns.py#L136 by adding `initial=status`. Then you can access the initial data inside the form via `self.initial`.

steglicd commented:
I agree with you about the two password fields. I will change this and will add a checkbox to display the password in plain text.

fonfon commented:
I'll review the pull request this week, probably on Thursday - sorry for the delay!

steglicd commented:
No problem. Currently I'm extending the plugin to offer a generic update URL feature.

steglicd commented:
@fonfon 
I added a first version of "Update URL"  feature. Even if GnuDIP is much more secure in my opinion, It is not well supported by most dynamic DNS providers. Update URL is a simple HTTP post of username/password/current IP to your Dynamic DNS provider and should work with most providers.

It's a "first version" because there is still a lot of work to do. I never used javascript bevore - so I surely need to improve that. Also I'm not happy with keeping the enumerated lists of the dropdown menu synchron between the template file and the python file. I will work on this. 

Also helper Texts and list of update URL templates are not complete yet.

And finally It would be nice to use plinth stylesheet files for the form (even if the plinth dropdown box looks a bit strange in my opinion). I will work on this too.

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,352 @@
+
+    enabled = forms.BooleanField(label=_('Enable Dynamic DNS'),
+                                 required=False, widget=forms.CheckboxInput
+                                (attrs={'onclick': 'mod_form();'}))
```
The 'onclick' event won't be triggered if you change the form with the keyboard space key. you could use jQuery 'change' instead

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,144 @@
+{% block page_js %}
+
+  <script type="text/javascript">
+  var SELFHOST = 'https://carol.selfhost.de/update?username=<Usere>&password=<Pass>&myip=<Ip>'
```
Usere (typo)

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,144 @@
+  <script type="text/javascript">
+  var SELFHOST = 'https://carol.selfhost.de/update?username=<Usere>&password=<Pass>&myip=<Ip>'
+  var NOIP = 'http://dynupdate.no-ip.com/nic/update?hostname=<Domain>&myip=<Ip>'
+  window.onload = function() {mod_form()}
```
I guess you could use 'mod_form()' directly also

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,352 @@
+                                             ('2', 'noip.com'),
+                                             ('3', 'selfhost.bz'),
+                                             ('4', 'other Update URL')),
+                                     widget=forms.Select(attrs={'onChange':
```
Don't put javascript into python files, but in the template or into a separate .js file if it gets a lot of code.

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,144 @@
+  var NOIP = 'http://dynupdate.no-ip.com/nic/update?hostname=<Domain>&myip=<Ip>'
+  window.onload = function() {mod_form()}
+
+	function show_pass()
```
You mix spaces and tabs here; For indentation in html/js files we use 2 spaces so far.

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,144 @@
+		if (service_type == 1){
+			document.getElementById('div-dynamicdns-updateurl').style.display = 'none';
+			document.getElementById('div-dynamicdns-gnudip').style.display = 'block';
+			document.getElementById('id_dynamicdns-dynamicdns_update_url').value='';
```
Is it really necessary to delete the value here? (just a thought; I don't know, didn't look into the details of the functionality)

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,352 @@
+    if new_status['dynamicdns_ipurl'] == '':
+        new_status['dynamicdns_ipurl'] = 'none'
+
+    if old_status['dynamicdns_server'] != \
```
Could you also compare the dictionaries directly with `if old_status == new_status`?

fonfon commented:
Thanks for the update! I focused on reviewing the code so far (instead of testing the functionality); A couple more thoughts/proposals:

- there's nothing wrong with plain javascript, but feel free to use jQuery where it makes your life easier
- forms layout: use the bootstrap layout instead of plain HTML, see forms of other modules for examples
- you could generate the whole form with `{{ form|bootstrap }}`. Do you have many customizations in there that wouldn't be covered by the generated form? The javascript would probably get a bit trickier but not too bad, i.e.:
```
      $('.form-group').hide();
      $('#id_dynamicdns-enabled').closest('.form-group').show();
```
- I get some flake8 errors about wrong indentation (using python3-flake8)

I'm available in the IRC channel today and happy to clarify things or help out! (JavaScript can be irritating in the beginning ;))

steglicd commented:
@fonfon as you proposed, I switched from plain javascript to jQuery, simplyfied argument comparsion, fixed indetion (whitespaces vs. tabs), removed javascript from python files, removed onclick events and some typo's.

I stil need to make the html files pep8 compatible. I will work on...

steglicd commented:
@fonfon I finished the points you mentioned. I added some update URL templates. The following services are now ready to use:
* gnudip (safe, free, but sadly not well known)
* noip (not free, but well known)
* selfhost.bz (free)
* freedns (free)

other services, based on an HTTP update, area able to be used via "other"

What I'm still not happy with:
syncing the dropdown box (strings and selected index) between python and htm template file

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns.html:
```diff
@@ -0,0 +1,38 @@
+   </br> </br>
+   If you are looking for a free dynamic DNS account, you may find one 
+   <a href='http://gnudip.datasystems24.net' target='_blank'> here</a>
+   or <a href='http://freedns.afraid.org/' target='_blank'> here</a></p>
```
the link names should be more descriptive

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,165 @@
+          $('#id_dynamicdns-dynamicdns_user').closest('.form-group').show();
+          $('#id_dynamicdns-dynamicdns_secret').closest('.form-group').show();
+          $('#id_dynamicdns-showpw').closest('.form-group').show();
+          $('#id_dynamicdns-dynamicdns_ipurl').closest('.form-group').show();
```
You could assign an additional class to the configuration form for clarity, say `dynamicdns-configuration` and try using `$('dynamicdns-configuration .form-group').show()`

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,165 @@
+
+      function show_all()
+      {
+          $('#id_dynamicdns-enabled').closest('.form-group').show();
```
the indentation is no the same as above

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,165 @@
+{% block page_js %}
+  <script type="text/javascript">
+
+    (function($) {
```
A general note:
It's good that you use a self-executing anonymous functions to not clutter the global namespace.
The functions in here (select_service etc.) are still global. You needed that earlier because some javascript function calls were in the html `attr`s, but it's not necessary anymore.

fonfon commented:
As you said the handling of the service type field could be better.
I was using an afraid.org account: The domain/username fields are obligatory although I don't need that for afraid.org (using the update URL). After submit - when asked to enter the missing values - my previous input for the update URL didn't show up anymore, I think that was some Javascript initialization. Also when disabling dynamicdns the form tells me to fill out these fields.

The form could be a bit cleaner and simpler. If I want to use the update URL I don't know which fields are actually relevant for me: Will HTTP basic auth be used, is this what the username/password is for? Or is this just for the GnuDIP configuration? So I wondered whether separating "update URL" and "GnuDIP" into different two forms makes sense. What do you think, are there too many fields used in both forms?

One more bug:
https://github.com/freedombox/Plinth/pull/81/files#diff-7d03a225ca7030e5d0ebe5b7123156d4R108
When displaying the form because not all fields were filled out, django pre-selects the submitted item. Then the text of this item is changed via javascript, and there are two options named "other update URL":
![screenshot - 2015-03-09 - 20 12 01](https://cloud.githubusercontent.com/assets/1867309/6562536/52977054-c69a-11e4-8990-860f289c7359.png)

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,165 @@
+                     'password=<Pass>&myip=<Ip>'
+      var NOIP     = 'http://dynupdate.no-ip.com/nic/update?hostname=' +
+                     '<Domain>&myip=<Ip>'
+      var FREEDNS  = 'http://freedns.afraid.org/dynamic/update.php?' +
```
https worked for me so it would be good to have it as default

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,372 @@
+                    &lt;Domain&gt; may be used within the URL. For details\
+                    see the update URL templates of the example providers.'
+
+    hlp_services = 'Please choose a update protocol according to your \
```
typo (an update)

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,394 @@
+
+    return TemplateResponse(request, 'dynamicdns.html',
+                            {'title': _('dynamicdns'),
+                             'subsubmenu': index_subsubmenu})
```
couldn't you use `subsubmenu` directly here and omit 
`index_subsubmenu = subsubmenu`?

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,394 @@
+        old_dynamicdns_secret = self.initial['dynamicdns_secret']
+
+        """clear the fields which are not in use"""
+        if service_type == '1':
```
The code would be easier understandable if you would use abbreviations instead of numbers for the select field (like 'gnudip', 'noip', ..)

fonfon commented on plinth/modules/dynamicdns/dynamicdns.py:
```diff
@@ -0,0 +1,394 @@
+    form = None
+
+    if request.method == 'POST':
+        form = ConfigureForm(request.POST, initial=status, prefix='dynamicdns')
```
There is only one form on this page so you could leave out the prefix, that would make statements like https://github.com/freedombox/Plinth/pull/81/files#diff-7d03a225ca7030e5d0ebe5b7123156d4R86 shorter (the `dynamicdns-dynamicdns` part)

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+      //show the enable checkbox
+      $('#id_dynamicdns-enabled').closest('.form-group').show();
+      if ($('#id_dynamicdns-enabled').prop('checked')) {
+      //show all form fields
```
indentation

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+        if ($('#id_dynamicdns-enabled').prop('checked')) {
+          show_all();
+          if ($("#id_dynamicdns-service_type option:selected").text() == "GnuDIP") {
+              set_gnudip_mode()
```
indentation (there are some more indentation mismatches further down)

fonfon commented:
Usually dynamicdns required the users to configure port forwarding on their router. We should mention that, in 'about' or so.

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+      }
+
+      function show_all()
+      {
```
instead of the lines here it'd be better to use `$('form.form .form-group').show();`
add a custom class to the form (i.e. `dynamicdns`) to make it a bit safer, then you can use `$('form.dynamicdns .form-group').show();`

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+        }
+      });
+
+      $('#id_dynamicdns-service_type').change(function() {
```
It would be better (faster and better readable) to get the service type value only once and reuse it then, like
`var service = $("#id_dynamicdns-service_type option:selected").text()`
and then `if (service == "noip.com") {`

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+            set_gnudip_mode()
+          }else{
+            set_update_url_mode();
+            if ($("#id_dynamicdns-service_type option:selected").text() == "noip.com") {
```
Your javascript syntax is inconsistent, you sometimes use spaces around brackets, sometimes not, sometimes the bracket is in the next line, sometimes not.
You could refer to some kind of style guide like https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml

fonfon commented on plinth/modules/dynamicdns/templates/dynamicdns_configure.html:
```diff
@@ -0,0 +1,168 @@
+
+  <form class="form" method="post">
+    {% csrf_token %}
+    <div id='dynamicdns-no-js'>
```
I'd use the official way here, the `<noscript>hej you have JS disabled!</noscript>` tag. Everything inside will only be displayed when javascript is disabled and you don't need to use additional javascript to hide it.

fonfon commented:
I didn't review the action script (https://github.com/steglicd/Plinth/blob/dynamicdns/actions/dynamicdns). If somebody reviews the script I'm okay with merging the pull request -- some things should be improved, but we can open issues for them or add inline TODOs:

- better synchronization of the service-type dropdown field between javascript and python (backend and frontend - again selecting the previously chosen service type)
- simplifying / usability of the configuration form: For example it's not clear which fields are used and required when choosing a service type. I thought about splitting GnuDIP from the url-based update mechanisms, but whatever makes the form easier to use is good.
- tidy up the javascript (always use `var`, always use semicolons)
- tidy up python: some of the functions perhaps can be implemented in a simpler / easier to read way, i.e. `_apply_changes` and `get_status`

I'm offline for 5 weeks now so I'll miss the next two ProgressCalls!

jvalleroy commented on actions/dynamicdns:
```diff
@@ -0,0 +1,434 @@
+        doGetWANIP
+        if [ "$(cat $HELPERCFG |grep ^NAT | awk '{print $2}')" = "no" ];then
+            #if we are not behind a NAT device and we use gnudip, start the daemon tool
+            if [ -f ${CFG} -a ! -z $(cat ${cfgfile} 2> /dev/null |grep server |cut -d = -f 2 |grep -v ^\'\')];then
```
shellcheck chokes on this line:
```
In actions/dynamicdns line 328:
            if [ -f ${CFG} -a ! -z $(cat ${cfgfile} 2> /dev/null |grep server |cut -d = -f 2 |grep -v ^\'\')];then
            ^-- SC1009: The mentioned parser error was in this if expression.
               ^-- SC1073: Couldn't parse this test expression.
                                                                                                             ^-- SC1020: You need a space before the ].
                                                                                                             ^-- SC1072: Unexpected ";". Fix any mentioned problems and try again.
```

steglicd commented:
@jvalleroy thanks for the feedack, I did not know shellcheck until now. Seems to be very useful.

I made the code shellcheck compatible.

steglicd commented:
@jvalleroy shellcheck gives me some errors:
SC2001
(means: I should avoid usage of sed and use built in string manipulation of bash).

But when I fix this, the script does not run under dash anymore. So I decided to ignore this error - if this is OK?

jvalleroy commented on actions/dynamicdns:
```diff
@@ -0,0 +1,443 @@
+            local wgetoptions=" --user ${user} --password ${pass} "
+        fi
+        if [ "${ignoreCertError}" = "enabled" ];then
+            local wgetoptions=" --no-check-certificate "
```
If wgetoptions was already defined on line 303, should the --no-check-certificate option be concatenated to the existing variable (instead of replacing it)?

jvalleroy commented:
@steglicd Yeah, I think we can ignore that error.

I added one more note about the script. Everything else looks ok to me.

steglicd commented:
@jvalleroy you are right! And also the global defined wget options (retry counter and stuff) should be used. I fixed this.

jvalleroy commented:
Ok, I've merged this into the master branch. We can open an issue to track any remaining TODOs like fonfon mentioned in https://github.com/freedombox/Plinth/pull/81#issuecomment-84199020.

Thanks for your contribution!
